mlir/lib/Dialect/XeGPU/Transforms/XeGPUUnroll.cpp

```diff
@@ -0,0 +1,466 @@
+      }
+    };
+
+    auto mixedOffsets = op.getMixedOffsets();
+
+    // For n-D memrefs where n > rank, we need to handle the last `rank`
+    // dimensions only, and keep the first `n-rank` dimensions as is.
+    SmallVector<OpFoldResult> oldOffsets = llvm::to_vector(
+        llvm::drop_begin(mixedOffsets, mixedOffsets.size() - rank));
+    auto validIdxes =
+        llvm::seq<int64_t>(mixedOffsets.size() - rank, mixedOffsets.size());
+
+    SmallVector<Value> newOps;
+    for (SmallVector<int64_t> offsets :
+         StaticTileOffsetRange(shape, targetShape)) {
+      for (auto [idx, oldOff, offset] :
+           llvm::zip(validIdxes, oldOffsets, offsets)) {
+        mixedOffsets[idx] = addi(oldOff, offset);
+      }
+      auto newOp = rewriter.create<xegpu::CreateNdDescOp>(
+          loc, newTdescTy, op.getSource(), mixedOffsets, op.getMixedSizes(),
+          op.getMixedStrides());
+      newOps.push_back(newOp);
+    }
+    auto castOp = unpack(newOps, tdescTy, targetShape, loc, rewriter);
+    rewriter.replaceOp(op, castOp);
+
+    return success();
+  }
+};
+
+struct UnrollUpdateNdOffsetOp : public UnrollPattern<xegpu::UpdateNdOffsetOp> {
+  using UnrollPattern<xegpu::UpdateNdOffsetOp>::UnrollPattern;
+  LogicalResult matchAndRewrite(xegpu::UpdateNdOffsetOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto tdesc = op.getTensorDesc();
+    auto tdescTy = tdesc.getType();
+    auto shape = tdescTy.getShape();
+
+    auto maybeTargetShape = getTargetShape(op);
+    if (!maybeTargetShape || llvm::equal(*maybeTargetShape, shape))
+      return failure();
+    auto targetShape = *maybeTargetShape;
+
+    auto convertedTdescTypes = getUnrolledTypes(tdescTy, targetShape);
+    auto convertedTdesc =
+        pack(tdesc, convertedTdescTypes, targetShape, loc, rewriter);
+
+    llvm::SmallVector<Value> newOps;
+    for (auto t : convertedTdesc) {
+      auto newOp = rewriter.create<xegpu::UpdateNdOffsetOp>(
+          loc, t.getType(), t, op.getOffsets(), op.getConstOffsets());
+      newOps.push_back(newOp);
+    }
+    auto castOp = unpack(newOps, op.getType(), targetShape, loc, rewriter);
+    rewriter.replaceOp(op, castOp);
+    return success();
+  }
+};
+
+struct UnrollPrefetchNdOp : public UnrollPattern<xegpu::PrefetchNdOp> {
+  using UnrollPattern<xegpu::PrefetchNdOp>::UnrollPattern;
+  LogicalResult matchAndRewrite(xegpu::PrefetchNdOp op,
+                                PatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto tdesc = op.getTensorDesc();
+    auto tdescTy = tdesc.getType();
+    auto shape = tdescTy.getShape();
+
+    auto maybeTargetShape = getTargetShape(op);
+    if (!maybeTargetShape || llvm::equal(*maybeTargetShape, shape))
+      return failure();
+    auto targetShape = *maybeTargetShape;
+
+    auto convertedTdescTypes = getUnrolledTypes(tdescTy, targetShape);
+    auto convertedTdesc =
```

Comment:
Use explicit type instead of auto: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

Reply:
Many place in this code might be more readable with explicit types.

Reply:
Thanks. Fixed.